packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol

```diff
@@ -340,23 +353,44 @@ abstract contract CrossDomainMessenger is
-        // Calldata overhead
-        + (uint64(_message.length) * MIN_GAS_CALLDATA_OVERHEAD)
-        // Dynamic overhead (EIP-150)
-        + ((_minGasLimit * MIN_GAS_DYNAMIC_OVERHEAD_NUMERATOR) / MIN_GAS_DYNAMIC_OVERHEAD_DENOMINATOR)
-        // Gas reserved for the worst-case cost of 3/5 of the `CALL` opcode's dynamic gas
-        // factors. (Conservative)
-        + RELAY_CALL_OVERHEAD
-        // Relay reserved gas (to ensure execution of `relayMessage` completes after the
-        // subcontext finishes executing) (Conservative)
-        + RELAY_RESERVED_GAS
-        // Gas reserved for the execution between the `hasMinGas` check and the `CALL`
-        // opcode. (Conservative)
-        + RELAY_GAS_CHECK_BUFFER;
+    function baseGas(bytes memory _message, uint32 _minGasLimit) public pure returns (uint64) {
+        // Base gas should really be computed on the fully encoded message but that would break the
+        // expected API, so we instead just add the encoding overhead to the message length inside
+        // of this function.
+
+        // We need a minimum amount of execution gas to ensure that the message will be received on
+        // the other side without running out of gas (stored within the failedMessages mapping).
+        // If we get beyond the hasMinGas check, then we *must* supply more than minGasLimit to
+        // the external call.
+        uint64 executionGas = uint64(
+            // Constant costs for relayMessage
+            RELAY_CONSTANT_OVERHEAD
+            // Covers dynamic parts of the CALL opcode
+            + RELAY_CALL_OVERHEAD
+            // Ensures execution of relayMessage completes after call
+            + RELAY_RESERVED_GAS
+            // Buffer between hasMinGas check and the CALL
+            + RELAY_GAS_CHECK_BUFFER
+            // Minimum gas limit, multiplied by 64/63 to account for EIP-150.
+            + ((_minGasLimit * MIN_GAS_DYNAMIC_OVERHEAD_NUMERATOR) / MIN_GAS_DYNAMIC_OVERHEAD_DENOMINATOR)
+        );
+
+        // Total message size is the result of properly ABI encoding the call to relayMessage.
+        // Since we only get the message data and not the rest of the calldata, we use the
+        // ENCODING_OVERHEAD constant to conservatively account for the remaining bytes.
+        uint64 totalMessageSize = uint64(_message.length + ENCODING_OVERHEAD);
+
+        // Finally, replicate the transaction cost formula as defined after EIP-7623. This is
+        // mostly relevant in the L1 -> L2 case because we need to be able to cover the intrinsic
+        // cost of the message but it doesn't hurt in the L2 -> L1 case. After EIP-7623, the cost
+        // of a transaction is floored by its calldata size.
+        return TX_BASE_GAS
+            + uint64(
+                Math.max(
+                    executionGas + (totalMessageSize * MIN_GAS_CALLDATA_OVERHEAD),
+                    (totalMessageSize * FLOOR_CALLDATA_OVERHEAD)
+                )
+            );
```

Comment:
Let's also add a code comment for explaining why we don't need the `isContractCreation` for completeness

Reply:
Done